File: c/src/Lista_vetor.c
#include <stdio.h>
#include <stdlib.h>
#include "Lista_vetor.h"

void lista_vetor_inicializar(ListaVetor *l) {
    l->tamanho = 0;
}

void lista_vetor_inserir(ListaVetor *l, int valor) {
    if (l->tamanho >= MAX) {
        printf("Lista cheia!\n");
        return;
    }
    l->dados[l->tamanho++] = valor;
}

static void heapify(int arr[], int n, int i) {
    int maior = i;
    int esq = 2 * i + 1;
    int dir = 2 * i + 2;

    if (esq < n && arr[esq] > arr[maior]) maior = esq;
    if (dir < n && arr[dir] > arr[maior]) maior = dir;

    if (maior != i) {
        int temp = arr[i];
        arr[i] = arr[maior];
        arr[maior] = temp;
        heapify(arr, n, maior);
    }
}

void heap_sort_lista_vetor(ListaVetor *l) {
    int n = l->tamanho;
    if (n <= 1) return;

    for (int i = n / 2 - 1; i >= 0; i--) {
        heapify(l->dados, n, i);
    }

    for (int i = n - 1; i > 0; i--) {
        int temp = l->dados[0];
        l->dados[0] = l->dados[i];
        l->dados[i] = temp;
        heapify(l->dados, i, 0);
    }
}

File: java/src/ListaVetor.java
public class ListaVetor {
    private int[] dados;
    private int tamanho;
    
    public ListaVetor(int capacidade) {
        dados = new int[capacidade];
        tamanho = 0;
    }
    
    public void inserir(int valor) {
        if (tamanho >= dados.length) {
            System.out.println("Lista cheia!");
            return;
        }
        dados[tamanho++] = valor;
    }
    
    public void heapSort() {
        if (tamanho <= 1) return;

        for (int i = tamanho / 2 - 1; i >= 0; i--) {
            heapify(dados, tamanho, i);
        }

        for (int i = tamanho - 1; i > 0; i--) {
            int swap = dados[0];
            dados[0] = dados[i];
            dados[i] = swap;
            
            heapify(dados, i, 0);
        }
    }
    
    private void heapify(int[] arr, int n, int i) {
        int maior = i;
        int esq = 2 * i + 1;
        int dir = 2 * i + 2;
        
        if (esq < n && arr[esq] > arr[maior]) {
            maior = esq;
        }
        
        if (dir < n && arr[dir] > arr[maior]) {
            maior = dir;
        }
        
        if (maior != i) {
            int swap = arr[i];
            arr[i] = arr[maior];
            arr[maior] = swap;
            
            heapify(arr, n, maior);
        }
    }
}